### PR TITLE
use readable-stream writable to correct res pipe

### PR DESCRIPTION
**Pull request by Matt-Esch**
I'm seeing interesting behavior with pipe into responses across various versions of node. In some versions I see empty data, and it seems really difficult to solve with an event-emitter based implementation only. The most compatible way to solve this problem is to inherit from readable-stream v2 writable, which I have done here.

I have also added an example to prove that proxying requests with request actually works, as this was my original breaking use case.

I have tested with all versions since 0.10. I would highly recommend releasing this as a patch version i.e. v3.0.1

Also as an interesting observation, by inheriting from writable I think this technically works with node 0.8, but the modules included for tests do not seem to work with node 0.8.

### CODE REVIEW

**Comment by doanythingfordethklok** [comment]
Can we drop support for very old versions of node? It seems that everyone should be migrated to an LTS version which has a large number of security fixes.

**Comment by Matt-Esch** [comment]
Other than deciding how to deal with headers, the streams support is now fully provided by the readable-stream module. readable-stream v2 supports all versions of node, so we don't have to do any heavy lifting. My concern for older version support is because I am still running 0.10 in production for a production service that's using this. You'd think we would have upgraded but there have been performance characteristics that prevented it, along with the mountain of subtly breaking things, including this module.

**Comment by doanythingfordethklok** [comment]
ok cool.  So I'll merge this and we'll update docs to only support node 0.10 or higher.  If things get weird anywhere else, we can have multiple implementations here then use `process.version` in `index.js` to switch which version is exported.

File: tests/pipe.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        .reply(200, '{a');
+
+    request.post('http://localhost:5555/derp', onResponse);
+
+    function onResponse(err, res, body) {
+        assert.ifError(err, 'Unexpected error in resonse');
+        assert.strictEqual('{a', String(body));
+        assert.end();
+    }
+});
```
Comment (doanythingfordethklok) [inline tests/pipe.js]: I'm not sure why the above test is here as it tests nock and request.

File: tests/pipe.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        var data = Buffer.concat(chunks);
+        assert.strictEqual(String(data), '{b');
+        assert.end();
+    });
+
+    request.post({
+        url: 'http://localhost:5555/derp',
+        body: '{a'
+    }).pipe(p);
+});
```
Comment (doanythingfordethklok) [inline tests/pipe.js]: I'm not sure why the above test is here as it tests nock and request.

File: tests/pipe.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        assert.strictEqual(String(data), '{b');
+        assert.end();
+    });
+
+    req.pipe(request.post('http://localhost:5555/derp'))
+        .pipe(p);
+
+    req.write('{a');
+    req.end();
+});
```
Comment (doanythingfordethklok) [inline tests/pipe.js]: It seems there is a better SSCCE by removing them and piping from MockRequest to PassThrough, then vrerifying on passthrough.  The request and nock modules seem irrelevant to whether MockRequest is working as expected..

File: tests/pipe.js (hunk 1 above)
Comment (Matt-Esch) [reply tests/pipe.js]: The point here was to prove that these modules actually work... What I wanted was a reproduction of a legitimate proxy case. However, when you have 3 different modules piped together and all you're looking at is what comes out of the pipe chain, it's not obvious which thing is broken. So really I just saw this as pre-requisite to the real proxy test through the hammock req/res pair.

File: tests/pipe.js (hunk 3 above)
Comment (Matt-Esch) [reply tests/pipe.js]: It's very subtle because request writes headers. PassThrough does not receive headers or set headers on the destination. Using passthrough would only verify the body.

Comment (doanythingfordethklok) [reply tests/pipe.js]: that is correct. streams are about piping data not doing http things like writing headers. Why would we expect pipe to write a header?

Comment (Matt-Esch) [reply tests/pipe.js]: Streams have multiple modes of operation (if you think about object mode, for example), but you could also ask why pipe doesn't send the full tcp payload and only the body section of the underlying tcp stream. I'm not sure I fully agree with this reasoning, but the reality is that the request module does indeed read headers from the source on request and set headers on the destination on response.

If you would rather not have tests for interop with request that's fine. I do think it's nice to have an example of a heavily used module that interacts with node req/res pairs and prove that the interaction with hammock doesn't deviate.

Comment (doanythingfordethklok) [reply tests/pipe.js]: The interop and integration tests are a good idea. 💯 I'd like to mark them as integration tests and have them be separate from unit tests that code against the node core interfaces.  (I admit this is a bit pedantic)

I made another branch and I'm also integrating travis ci here since it will make things better.

Comment (doanythingfordethklok) [reply tests/pipe.js]: as for `stream.pipe()` sometimes setting headers, I don't really think that was a good idea as it mixes the stream responsibilities with the http ones.. 

and yeah, we cannot change that so let's just work with it.

File: tests/pipe.js (hunk 1 above)
Comment (doanythingfordethklok) [reply tests/pipe.js]: that makes sense. 💯  I suggest we move this into a test that is separate from the unit test.

**Comment by doanythingfordethklok** [comment]
updated test layouts and added travis ci to PR 35
